### PR TITLE
Move "opinionated" Gutenberg block styles to theme.scss

### LINKED ISSUE #7172: Move opinionated styles to theme.css
The addition of an opt in theme.css is a good thing - but it's not currently useful. There's still loads of content in styles.css that is opinionated and not structural.

I think that theme.css should contain anything that changes how the content looks. Color, background color, margins & padding (with sizes - auto for centering is fine), font sizes, borders. The only things that should be loaded globally are things that affect position/ layout.

Some examples:

`p.wp-block-subhead{font-size:1.1em;font-style:italic;opacity:.75}`
The font size doesn't fit my typographic scale. The style and opacity break my styleguide.

`.wp-block-pullquote{border-top:4px solid #555d66;border-bottom:4px solid #555d66;color:#40464d;padding:3em 0;text-align:center}`
The borders and colours don't match the themes quote styles. The color doesn't fit in my colour palette work, and won't show on themes with a dark background. The padding breaks the themes vertical rhythm. The centering is ok I guess, but is also opinionated. All of these styles are attached to blockquotes which I already style so I want to turn them off entirely.

`.wp-block-image.is-resized figcaption{-ms-grid-row:2}`
May need a separate ticket for this but why is there no grid-row rule here? And why no other prefixes?

At the moment I am considering dequeuing all gutenberg styles from the front and back end using my own since it's more effort to override default styles than it is to write it again from scratch.

My other thought is to include the block .scss files, and then I can only use the ones I want - but that may not be so straight forward once the code is in core.

### PR DESCRIPTION
**Pull request by jasmussen**
**Goal:** With 3.0, we merged in changes to allow Gutenberg to provide _opinionated styles_ for blocks. For example, to provide CSS styles for the `blockquote`, which is traditionally styled by the theme itself. Simply put, a `theme.scss` exists for each block, and a theme can opt out of those styles, indicating a desire to provide these styles itself. See also https://wordpress.org/gutenberg/handbook/extensibility/theme-support/#default-block-styles.  Fixes #7172.

This moves opinionated styles, such as colors, fonts, and other non-structural elements to separate theme.scss files which require opt in by the theme itself. This includes in this PR:

- Code styles
- Embed captions
- Image captions
- Preformatted styles
- Pullquote fonts & border styles
- Quote styles
- Table styles
- Video captions

In addition to this there's a new caption style mixin which reduces caption style reuse.

### CODE REVIEW

**Comment by BinaryMoon** [comment]
You opt into the theme styles with `add_theme_support( 'wp-block-styles' );` - this is in the Gutenberg starter theme.

In the editor style.scss, edit.scss, and theme.scss are all used; however I believe (hope) you can dequeue them individually.

In the theme only style.scss is used by default, and theme.scss is opt in.

**Comment by BinaryMoon** [comment]
> Given Pullquote is a new block, do styles here need to be opt-out-able, or are you automatically opting in given it's a new block?

It depends on the markup.

Do pullquotes use alignleft and alignright? Those are required items for themes on wordpress.org so themes should already account for a lot of the positional styles. You should then move the visual styles (borders and font sizes) into theme.scss so that they don't conflict with theme defaults.

> Related to that, given the recent improvements to block variations (#7362), do we retire the Pullquote block and make it a variation of the Quote block? (#5947)

Personally I would prefer that pull quote gets removed. I'm not sure the average user will know the difference between a block quote and a pull quote.

> There are a lot of duplicated caption styles that are all the same. Is there a refactor we can do here?
Is "preformatted" a block we want to provide opinionated styles on?

I haven't looked at the code but I think consistency is super important.

> How about subheading?

How does a subheader differ from a heading set to h3/ h4? I don't know the answer and I'm a web designer. How will a non-webdesigner know the answer?

> Table we almost certainly want to provide some opinionated styles on. I can imagine many themes wanting to opt into these as it can be a slog to duplicate the table code from theme to theme, and it would be nice to just let Gutenberg do this. Can a theme opt in or out on a per-block basis?

AFAIK Only opt in for everything currently. What styles need copying across?

What styles should be opinionated on tables? Do themes not include table styles normally (mine do)?

> Videos and video embeds in Gutenberg are responsive. Should this responsiveness live in theme.scss?

I currently use jetpack responsive videos. Is this no longer needed? I know it's a separate topic but as a user I find it confusing that there's so many video embeds - I keep adding the video block and then remebering I have to add the Youtube block. I miss just being able to paste a video url and have it work.

> Other blocks to discuss include table, text columns, verse, video

Text columns should be included in style.scss since they aren't currently available.

**Comment by jasmussen** [comment]
Solid feedback, Ben, thank you. 

> the way I think of it, the styles should be in theme.css if they are styles that would already be styled in most existing themes. 

I think that's a good rule of thumb, and important to balance this given back compat with themes. This is why I'm asking on a per-block basis. I don't have strong opinions here myself, but want to make sure that any benefits Gutenberg can offer are offered.

For example, I totally agree that the stock `blockquote` is theme territory unless the theme specifically says "sure, lay Gutenberg styles on me". However we might want to offer the additional quote variations even without opt-in — otherwise a user might pick the 2nd ("Large") quote style only to see no change on the front-end. 

Similarly, in Gutenberg an image that is not inserted as an inline image comes with new `figure` markup and accompanying `figcaptions`. Do we offer those caption styles by default or should they require opt-in? 

> Personally I would prefer that pull quote gets removed. I'm not sure the average user will know the difference between a block quote and a pull quote.

I agree, and this circles back to quote style variations again — a pullquote style should probably have at least some structural visuals offered, like a font size, even if a theme hasn't necessarily opted in.

> How does a subheader differ from a heading set to h3/ h4? I don't know the answer and I'm a web designer. How will a non-webdesigner know the answer?

The subheading block is a weirdly controversial one. I don't quite understand its purpose myself, and it seems half want it removed, while for the other half it's a make or break block. I don't have strong opinions here myself. There's a ticket somewhere discussing this. I read it as "Dr. Strangelove" being the heading, and "Or: How I Learned to Stop Worrying and Love the Bomb" as the subheading.

> AFAIK Only opt in for everything currently. What styles need copying across?

No strong opinions here. I was thinking border-width and border-collapsing. I tried to make it responsive but failed, this might come in a separate block.

> I currently use jetpack responsive videos. Is this no longer needed? 

Possibly?

embeds

>  I miss just being able to paste a video url and have it work.

Is this not working for you? It's working fine for me. Try pasting `https://www.youtube.com/watch?v=tEJpLDEOivA` on a new line. The Video block is for videos you upload yourself, or direct links to video URLs. The embed blocks are just aliases for every existing WordPress supported oEmbed. 

Thanks again Ben, I hope to address all your thoughts in this PR — _possibly_ today, otherwise next week. We should be able to get these in 3.3. Feel free to lay other concerns on me, if there are frustrations I can address, you know where to find me.

**Comment by jasmussen** [comment]
Okay @BinaryMoon, I took a stab at your feedback. Please take a look at the diff and see if you feel it addresses your thoughts. Thank you.

https://github.com/WordPress/gutenberg/pull/7624/commits/57787fd79b7306e84bc5b11536895764b19aaf44

**Comment by samikeijonen** [comment]
@BinaryMoon I tried to dequeue `theme.css` but couldn't do it with `wp_dequeue_style( 'wp-core-blocks-theme' );`

Tried `enqueue_block_editor_assets` and `enqueue_block_assets` hooks. Let me know if you already know how to do it.

**Comment by jasmussen** [comment]
@samikeijonen If you simply _don't_ add `add_theme_support( 'wp-block-styles' );` to your theme `functions.php`, no "opinionated" styles should be loaded for you. Especially after this PR gets merged, I suppose — it's still early days in `master`. See also https://wordpress.org/gutenberg/handbook/extensibility/theme-support/#default-block-styles

**Comment by jasmussen** [comment]
Ticket is thoroughly squashed and rebased since the first attempt. So the diff should be easy to follow. I've also renamed the ticket to represent that this is no longer an experiment.

**Comment by samikeijonen** [comment]
@jasmussen `theme.css` is always loading in the editor. That's what I try to dequeue.

**Comment by jasmussen** [comment]
Ah, understood. Yes it seems like if you intend to provide your own styles for both places, it could make sense to let you control that. However it seems like whereas these styles should be _opt-in_ by the theme, they should be _opt-out_ by the editor, right?

**Comment by jasmussen** [comment]
@BinaryMoon Ben is there any chance you can take a look at this PR again? I know you're a busy man, but I would very much love to make the next version of Gutenberg easier for you to style. ❤️

**Comment by BinaryMoon** [comment]
@jasmussen - Hi Joen - I didn't realise you were waiting for me. Sorry about that.

I think the changes you have suggested help quite a bit, but they're not perfect.

* code - looks good
* embed - needs the figcaption removed from styles.scss - it's currently in both theme.scss and styles.scss
* gallery - should the figcaption use the caption mixin? Plus you should remove the duplicate styles from the gallery style.scss. They should be in style.scss or theme.scss - not both.
* image - perhaps not the right place to ask this but why is max-width set to none here? This breaks responsive behaviours.
* preformatted - probably don't need the white space rule since themes should include that, but I guess it doesn't matter. Otherwise looks good.
* pullquote - why are line height and font size being set for paragraphs? Shouldn't they inherit the themes styles?
* quote - I think everything from quote should be in theme.scss (as per my pull request). Blockquotes should already have margins and padding set, and font styles should also follow theme standards. Regarding the text-align:right in the quote footer... does Gutenberg take care of rtl styles?
* table - looks good

**Comment by jasmussen** [comment]
Thank you for the review again. I'm not waiting on you, I just appreciate your feedback as I know how deep into this you are.

I will take another stab at this and fix as soon as I can.

**Comment by jasmussen** [comment]
Thanks Ben. Pushed some changes:

- Removed figcaption from embed.
- Removed the theme.scss for the gallery entirely. Since it's a brand new block with unique markup, arguably all the CSS is structural.

To your questions:

> should the figcaption use the caption mixin

No, because it's sort of a unique overlay caption style here. 

> perhaps not the right place to ask this but why is max-width set to none here? This breaks responsive behaviours.

The history of that decision lives in #6496. But the short of it is that when an image is resized, we use the figure element to define the width of the element so that the figcaption could have the same width as the image. However I see your point, and I will make a note to revisit this, though I'd like to do that separately. 

> preformatted - probably don't need the white space rule since themes should include that, but I guess it doesn't matter. Otherwise looks good.

Not entirely sure I got this, as the rule is primarily for use within the editor. Were you suggesting this rule should've been in style or theme instead?

> pullquote - why are line height and font size being set for paragraphs? Shouldn't they inherit the themes styles?

No strong opinion here, but given this is an entirely new block (and by the way possibly one that will be removed and turned into a quote variation instead), I felt that this needed some styling in order to be a separate block at all. In this case, the font size of the body text is rather large, even if it's a stock p tag. Given that this block will likely transform into being a variation of the quote (see https://github.com/WordPress/gutenberg/issues/5947), okay to skip on this one for now?

> quote - I think everything from quote should be in theme.scss (as per my pull request). Blockquotes should already have margins and padding set, and font styles should also follow theme standards. Regarding the text-align:right in the quote footer... does Gutenberg take care of rtl styles?

As is in this branch right now, every stock standard blockquote is styled by the theme. All that's left in style.css is for the "large" variation of the quote:

screen shot 2018-07-04 at 13 40 16

And even then, only the font sizes to ensure it's a variation at all.

This touches sort of on an issue here — I completely understand why you as a theme developer would not want this, because I trust you to provide your own CSS for large variations. 

However for the millions of existing themes out there that do not provide these classes, a user will use Gutenberg to use a large variation of the quote style, and not see any change reflected in the content. What's the best solution here? 

For example, the separator block also just received variations:

screen shot 2018-07-04 at 13 43 30

Personally, I feel like it's a good trade-off for Gutenberg to leave the default separator and default blockquote alone, unless a theme explicitly opts into this, but leaving structural CSS for any variations offered. Or should this aspect be discussed as a separate issue?

**Comment by BinaryMoon** [comment]
Hey Joen - this is looking a lot better to me.

* embed 👍
* gallery 👍

> The history of that decision lives in #6496. But the short of it is that when an image is resized, we use the figure element to define the width of the element so that the figcaption could have the same width as the image. However I see your point, and I will make a note to revisit this, though I'd like to do that separately.

I ran into this being a problem using Gutenberg on my personal site on [this post](https://www.binarymoon.co.uk/2018/06/brush-ninja/). The animations in the sidebar were overlapping the content and wouldn't scale. Happy to offer feedback/ testing when you create a new ticket.

> (Re: preformatted) Not entirely sure I got this, as the rule is primarily for use within the editor. Were you suggesting this rule should've been in style or theme instead?

My bad. I thought the code was in theme.scss - I misread the filename.

> (Re: pullquote) No strong opinion here, but given this is an entirely new block (and by the way possibly one that will be removed and turned into a quote variation instead), I felt that this needed some styling in order to be a separate block at all. In this case, the font size of the body text is rather large, even if it's a stock p tag. Given that this block will likely transform into being a variation of the quote (see #5947), okay to skip on this one for now?

I'm all for removing pullquote so lets ignore for now.

> This touches sort of on an issue here — I completely understand why you as a theme developer would not want this, because I trust you to provide your own CSS for large variations.

That is a challenging one and I totally see your point. The variations is harder to style with default themes.

I wonder if the core code should be changed so that the default variants are only enabled when the default styles are opted in to? That way you can include all the variants in styles.scss and keep theme.scss clean for the rest of us (and we can include our own variants within the themes).

Although that would then raise issues with portability & lock in.

I hadn't seen the separator variations. That's an interesting addition. I think having variants is good - but I can't decide if having quite so many is a good thing. I can imagine this being abused.

**Comment by jasmussen** [comment]
Given the variations are in master, and we will be following up on this in the pullquote ticket, I would think it good to get this branch in as soon as we can, and then iterate further separately. Alternately this is likely to land in the 3.3 branch instead of 3.2.

**Comment by BinaryMoon** [comment]
Let's merge then! It's definitely a step in the right direction.

**Comment by jasmussen** [comment]
Just tested this again, it seems solid and harmless to me. Awaiting @karmatosed call on whether this can make the 3.2 deadline. 🤞

File: core-blocks/code/theme.scss
Comment (aduth) [inline core-blocks/code/theme.scss]: Next person to edit this file, if they're using the provided EditorConfig, will introduce a trailing new line to the file.

I'd suggest installing a plugin for your editor so we can be consistent on this:

https://editorconfig.org/#download

Comment (jasmussen) [reply core-blocks/code/theme.scss]: Found it. Turns out that's a VSCode setting that defaults to off, which confused me because I switched from Atom where it defaults to on:

screen shot 2018-07-06 at 15 27 41

I've now fixed this, and made a backup of my user config. Sorry about that.

**Review by karmatosed — APPROVED** [review]
Looks great to me let's 🚢

**Pull request merged**
